crates/common/src/path.rs

```diff
@@ -0,0 +1,47 @@
+use std::io;
+
+use normpath::error::ParentError;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum PathsError {
+    #[error("Cannot get root directory")]
+    IO {
+        #[from]
+        source: io::Error,
+    },
+    #[error("Cannot get root directory as parent of path")]
+    Parent(#[from] ParentError),
+    #[error("Cannot get project root, parent missing")]
+    NoParent,
+}
+
+#[macro_export]
+macro_rules! make_path {
+    ($first:expr, $($segments:expr),+) => {{
+        let mut path = common::PathBuf::new($first).expect("Invalid base path!");
+        $(path.push($segments);)*
+
+        path = {
+            if let Ok(normalized) = path.normalize() {
+                normalized
+            } else {
+                path.normalize_virtually().expect("Invalid path!")
+            }
+        };
+
+        // Automatically create directories to avoid errors and improve discoverability
+        // Don't try to create a directory named scc.exe
+        if path.extension().is_none() {
+            let _ = std::fs::create_dir_all(&path);
+        } else {
+            if let Ok(base_parent) = path.parent() {
+                if let Some(parent) = base_parent {
+                    let _ = std::fs::create_dir_all(&parent);
+                }
+            }
+        }
+
+        path
+    }}
+}
```

Comment:
any reason why this isn't a function?
should probably be something like
```rs
fn make_path(arg: impl IntoIterator<Item=impl AsRef<Path>>) -> ...
```

Reply:
Well, my understanding is that:
```rust
fn make_path(arg: impl IntoIterator<Item=impl AsRef<Path>>) -> ...
```
is syntactic sugar for:
```rust
pub fn make_path<I, P>(segments: I) -> Result<PathBuf, ErrorPathDoesNotExist>
where
    I: IntoIterator<Item = P>,
    P: AsRef<std::path::Path>,
...
```

Which would make calls like: `make_path([&root, "target", "staging"])` invalid.

While each specific item in that call satisfies the bound `AsRef<std::path::Path>`, all instances of a generic parameter have to be the same type. So the bound type `P` becomes `&PathBuf` and then `"target"` causes a type mismatch error.

I believe this is part of the reason `std::fmt` presents functionality like `println` as macros, not as functions.

Reply:
If you want to use a macro for that you should extract just that bit, e.g.
```
macro_rules! make_path {
    ($first:expr, $($segments:expr),+) => {{
        let mut path = common::PathBuf::new($first).expect("Invalid base path!");
        $(path.push($segments);)*
        path
    }}
}
```
instead of having everything in the macro

Reply:
another minor note, but for good macro hygiene it's usually considered good practice to use explicit global scope for paths: `::common::PathBuf::new` instead of `common::PathBuf::new`

Reply:
> use explicit global scope for paths: `::common::PathBuf::new` instead of `common::PathBuf::new`

Ah, I didn't know about the `::` prefix syntax. I thought I was already using global scope what with relative specifiers like `self::`, `super::` and `crate::` existing. Good to know, thanks!

Reply:
the `::` bit is nice for macros, because the user could define their own `common` module and it'd lead to an ambiguity since macros are expanded wherever they're used; you don't need to worry about it outside of macros, but it's worth it to be careful in macros

Reply:
> If you want to use a macro for that you should extract just that bit, e.g. `...` instead of having everything in the macro

So, you are saying that:
```rust
    let staging = make_path(make_path_join_macro!(&root, "target", "staging"));
```
instead of:
```rust
    let staging = make_path!(&root, "target", "staging");
```
would be the preferred way to call it?

Reply:
you can extract the bit that does all the other stuff to a function and call it from the macro, most of the code inside the macro is monomorphic and could be put in a function (which will also make the life a bit easier for the compiler since it'll have less parsing and compiling to do)

Reply:
the problem with big macros is they can easily explode code size, since it's basically just copy pasting tokens